Yoshiurr/pyomoexpression

yoshiurr-INL opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address?

Issue #42

##### What are the significant changes in functionality due to this change request?

When triggered, TEAL will accept pyomo expression as an acceptable format and vprints are changed accordingly. In that case, the output should be equation expressions instead of float values.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/HERON/wiki/Code-Standards) for details.
- [ ] 4. Automated Tests should pass.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large tes.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added, the the analytic documentation must be updated/added.
- [x] 9. If any test used as a basis for documentation examples have been changed, the associated documentation must be reviewed and assured the text matches the example.

yoshiurr-INL committed:
Update main.py

Updated for pyomo expressions to pass without errors.

yoshiurr-INL committed:
Update CashFlows.py

Updated to pass pyomo expressions without errors.

yoshiurr-INL committed:
Add files via upload

New python object oriented file for hourly NPV with pyomo expressions.

yoshiurr-INL committed:
Update HourlyObjectOrientedTestWithPyomoExpression.py

Updated file.

yoshiurr-INL committed:
Update CashFlows.py

Updated file.

yoshiurr-INL committed:
Update main.py

Updated File

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -971,14 +971,18 @@ def checkInitialization(self):
     """
       Initialize some parameters
       @ In, lifetime, int, the given life time
       @ Out, None
```
this docstring needs to be updated with the optional variable addition

PaulTalbot-INL replied:
it looks like `pyomoVar` here can be a boolean that defaults to `False`

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -1011,10 +1015,15 @@ def extendParameters(self, toExtend, t):
     for name, value in toExtend.items():
       if name.lower() in ['alpha', 'driver']:
-        if utils.isAFloatOrInt(value) or (len(value) == 1 and utils.isAFloatOrInt(value[0])):
+        if "pyomo.core.expr" in str(type(value)):
```
I'm still not sure what the best way to act on this is; I really dislike string-type checking, but I don't really want to add pyomo as a dependency to TEAL, either.

PaulTalbot-INL replied:
I found some interesting options, but it occurred to me the option may be even easier: can we check the opposite? i.e. if it's a float (or array of floats), then treat it as such; else treat it like pyomo (or anything else, since it uses "object")

yoshiurr-INL replied:
That is a nearly perfect answer! If further distinction is required later, then we'll leave it to future PRs.

yoshiurr-INL replied:
I hit an unexpected error. Apparently, len() for a pyomo expressionis not valid. The evaluation 'len(value) == 1 ' is triggering the error. Would it be acceptable to remove this condition and afterwards check if the length is one?
Like:
if utils.isAFloatOrInt(value):
  if (len(value) == 1 and utils.isAFloatOrInt(value[0])):
    ........

yoshiurr-INL replied:
Would this be doubly checking too much?

PaulTalbot-INL replied:
Hm, interesting. We should probably do a type check first and then only check length if it's a np.array or float type; otherwise we treat it as a pyomo object. Data checking here might get tricky; let me know and we can look at it together if useful.

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -1110,14 +1119,18 @@ def __init__(self, **kwargs):
     """
       Initialize some parameters
       @ In, lifetime, int, the given life time
+      @ In, pyomoVar, pyo.ConcreteModel(), concrete model including all variables
```
Do we actually want the ConcreteModel instance passed in, or should this be a boolean flag? Also, need to add "optional" after the type entry in this docstring.

yoshiurr-INL replied:
Yes, it should be boolean now. The previous object array creation was by John's method and it required the initial concrete model. Now that it does not, it can be False/True boolean style.

PaulTalbot-INL commented on src/main.py:
```diff
@@ -190,13 +190,14 @@ def _createEvalProcess(components, variables):
     @ In, cf, CashFlows.CashFlow, cashflow who is being analyzed
     @ In, variables, dict, RAVEN variables as name: value
     @ In, v, int, verbosity
+    @ In, pyomoSwitch, pyo.ConcreteModel(), concrete model including all variables
```
can this be a boolean value that defaults to `False` or do we need the model?

Also, docstring needs the "optional" tag.

yoshiurr-INL replied:
Will change type and docstring for all similar comments below.

PaulTalbot-INL commented on src/main.py:
```diff
@@ -344,12 +361,20 @@ def projectSingleCashflow(cf, start, end, life, lifeCf, taxMult, inflRate, proje
     @ In, inflRate, float, inflation rate multiplier (1 - inflation)
     @ In, projectLength, int, total years of analysis
     @ In, v, int, verbosity
+    @ In, pyomoSing, pyo.ConcreteModel(), concrete model including all variables
```
as with the others, can we get by with a boolean check here? also optional tag in docstring

PaulTalbot-INL commented on src/main.py:
```diff
@@ -423,46 +451,58 @@ def npvSearch(settings, components, cashFlows, projectLength, v=100):
     @ In, projectLength, int, project years
     @ In, mult, float, optional, if provided then scale target cash flow by value
     @ In, v, int, verbosity level
+    @ In, pyomoFCFF, pyo.ConcreteModel(), concrete model including all variables
```
boolean?

PaulTalbot-INL commented on src/main.py:
```diff
@@ -423,46 +451,58 @@ def npvSearch(settings, components, cashFlows, projectLength, v=100):
     @ In, projectLength, int, project years
     @ In, discountRate, float, firm discount rate to use in discounting future dollars value
     @ In, mult, float, optional, if provided then scale target cash flow by value
+    @ In, pyomoNPV, pyo.ConcreteModel(), concrete model including all variables
```
boolean?

PaulTalbot-INL commented on src/main.py:
```diff
@@ -536,11 +576,12 @@ def lcmm(*args):
 # MAIN METHOD
 #=====================
-def run(settings, components, variables):
+def run(settings, components, variables, pyomoChk=None):
```
boolean?

PaulTalbot-INL commented on tests/HourlyObjectOrientedTestWithPyomoExpression.py:
```diff
@@ -0,0 +1,276 @@
+  dictDf['B'] = B
+  dictDf['C'] = C
+  dictDf['D'] = D
+  dfSet = (df, dictDf)
```
we need more clarity, perhaps from comments or variable naming, to explain what data is coming from the CSV and what data is pyomo vardata. Where possible, we should eliminate the interaction with the CSV.

yoshiurr-INL replied:
Thank you for leaving the comments above! Knowing your time is extremely limited, I appreciate the guidance and assistance!

PaulTalbot-INL reviewed (changes requested):
@yoshiurr-INL I know you're still tinkering with this, but added a few review comments to consider.

We definitely need to solidify what ConcreteModel optimization is being performed in the test example; we can probably make use of one of the examples in HERON with a couple components adding technical constraints, and using the metric as an optimization target.

PaulTalbot-INL commented:
@worseliz this is the PR looking at adding pyomo-compatible cash flow analysis. The need comes from the DISPATCHES project requiring TEAL as a source for CashFlow expression constructions in their pyomo environment. We can have a chat about details if you want.

yoshiurr-INL committed:
Update HourlyObjectOrientedTestWithPyomoExpression.py

Updated based on comments provided.

yoshiurr-INL committed:
Update CashFlows.py

Updated based on comments provided.

yoshiurr-INL committed:
Update main.py

Updated based on provided comments.

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -1015,14 +1016,16 @@ def extendParameters(self, toExtend, t):
     for name, value in toExtend.items():
       if name.lower() in ['alpha', 'driver']:
-        if "pyomo.core.expr" in str(type(value)):
+        if utils.isAFloatOrInt(value):
```
does this util catch numpy arrays of floats and ints as well?

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -1119,15 +1122,15 @@ def __init__(self, **kwargs):
       Initialize some parameters
       @ In, lifetime, int, the given life time
-      @ In, pyomoVar, pyo.ConcreteModel(), concrete model including all variables
+      @ In, pyomoVar, boolean, 'True' will trigger pyomo flag
```
can we clarify this a bit for future developers, something like "If True, indicates that an expression will be constructed instead of a value calculated."

PaulTalbot-INL commented on src/main.py:
```diff
@@ -95,28 +95,30 @@ def checkRunSettings(settings, components):
     @ In, components, list, list of CashFlows.Component instances
     @ In, variables, dict, variable-value map from RAVEN
     @ In, v, int, verbosity level
+    @ In, pyomoDriv, boolean, 'True' will trigger pyomo flag
```
similar comment here, let's explain what the flag means for the sake of future dev

PaulTalbot-INL replied:
I think this is an outstanding need still.

PaulTalbot-INL commented on src/main.py:
```diff
@@ -95,28 +95,30 @@ def checkRunSettings(settings, components):
     Sorts the cashflow evaluation process so sensible evaluation order is used
     @ In, components, list, list of CashFlows.Component instances
     @ In, variables, dict, variable-value map from RAVEN
+    @ In, pyomoEval, boolean, 'True' will trigger pyomo flag
```
flag clarification

PaulTalbot-INL commented on src/main.py:
```diff
@@ -222,7 +224,7 @@ def componentLifeCashflow(comp, cf, variables, lifetimeCashflows, v=100, pyomoSw
         else:
           name = '(from input)'
-        if pyomoSwitch == None:
+        if pyomoSwitch == False:
```
since it's boolean, it's more standard to say `if pyomoSwitch:` than check equality to False.

PaulTalbot-INL replied:
this shows up a few times in these changes

yoshiurr-INL replied:
Do you mean:
if pyomoSwitch:
  if pyomoSwith == False:
    ...........

PaulTalbot-INL replied:
Given:
```python
myVar = False
```
then
```python
if not myVar:
  print('no equals signs')
if myVar == False:
 print('with equals signs')
```
both print out, and the former is preferred (no equals signs). `None` is slightly different, in that case the preferred check is
```python
myVar = None

if myVar is None:
  print('is none')
```
see https://www.python.org/dev/peps/pep-0008/

PaulTalbot-INL commented on src/main.py:
```diff
@@ -401,17 +403,22 @@ def projectSingleCashflow(cf, start, end, life, lifeCf, taxMult, inflRate, proje
-  #projCf[newBuildMask] = lifeCf[0] * taxMult * np.power(inflRate, -1*years[newBuildMask])
-  for i in range(len(newBuildMask[0])):
-    projCf[newBuildMask[0][i]] = lifeCf[0] * taxMult * np.power(inflRate, -1*years[newBuildMask[0][i]])
+  if pyomoSing == False:
```
`if not pyomoSing:`

PaulTalbot-INL commented on src/main.py:
```diff
@@ -591,7 +601,7 @@ def run(settings, components, variables, pyomoChk=None):
   # check mapping of drivers and determine order in which they should be evaluated
   vprint(v, 0, m, '... Checking if all drivers present ...')
-  ordered = checkDrivers(settings, components, variables, v=v)
+  ordered = checkDrivers(settings, components, variables, v=v, pyomoDriv=pyomoChk)
```
I noticed this flag changes names a few times. Would it be useful to make it consistent so that maintaining this feature is easier in the future? Or do the different flags do different things that I'm not seeing?

yoshiurr-INL replied:
I have it changing names just in case if someone changes the code in the future, the location of the error related to pyomo flags would be obvious.

PaulTalbot-INL replied:
The stack trace is python is pretty good at telling you which line an error is raised on; I would trust to that rather than using lots of renaming, as this obfuscates the meaning from future developers.

yoshiurr-INL committed:
Update CashFlows.py

Found compatibility issues when value in extendParameters is type np.array. Fixed issue.

yoshiurr-INL committed:
Update HourlyObjectOrientedTestWithPyomoExpression.py

An 's' was added in the run function by mistake. Fixed.

yoshiurr-INL committed:
Update main.py

Updated file according to pushed changes made to devel 3 days ago.

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -1016,16 +1016,25 @@ def extendParameters(self, toExtend, t):
+            listArray = [0]*t
+            listArray[0] = value
+            toExtend[name] = np.array(listArray)
+        except:
```
all-catching "except"s are bad practice; if we can't check for float or numpy types, we need to be very specific about which exception we're catching.

PaulTalbot-INL commented on src/main.py:
```diff
@@ -652,39 +644,23 @@ def run(settings, components, variables, pyomoChk=False):
 
 
-  some_data = {**projectCashflows, **compCashflows }
+  results = {}
```
did you rebase or merge devel into this branch, or did you make changes manually?

yoshiurr-INL committed:
Merged with TEAL devel and updated extendParameters except the Amortizor

PaulTalbot-INL commented on src/main.py:
```diff
@@ -282,11 +301,17 @@ def projectLifeCashflows(settings, components, lifetimeCashflows, projectLength,
+  #   tax = comp.getTax() if comp.getTax() is not None else settings.getTax()
+  #   inflation = comp.getInflation() if comp.getInflation() is not None else settings.getInflation()
+  #   compProjCashflows = projectComponentCashflows(comp, tax, inflation, lifetimeCashflows[comp.name], projectLength, v=v)
+  #   compCashflows[comp.name] = compProjCashflows
```
this commented block should be removed

yoshiurr-INL committed:
Updated extendParameters once more since type str were being sent as pyomo inputs

yoshiurr-INL committed:
Solved issue of TEAL tests failing

PaulTalbot-INL commented on src/CashFlows.py:
```diff
@@ -1027,10 +1027,14 @@ def extendParameters(self, toExtend, t):
               new = np.zeros(t)
               new[0] = float(value)
               toExtend[name] = new
+            elif type(value) is str:
```
when do we get a string? Does this happen in real expected pyomo use or is this an artifact of the test used?

yoshiurr-INL replied:
This type is passed from 12 out of the 14 tests from TEAL tests. The reason they were failing earlier yesterday is because string type data was being passed to extendParameters in CashFlows.py. When that happens, the original condition of "utils.isIntorFloat(value) or (len(value)==1 and utils.isIntorFloat(value[0]))" lets string type alpha/driver to pass without being extended.

Earlier yesterday's code checked the "else" case and extended it regardless of its type. To accommodate, I have added the case of "if type(value) is str -> pass." 

Now it would be great to use "if type(value) == pyomo.blah.blah...." to avoid coincidental extension of non-int-float-str data types being passed as objects but there is no guarantee the type of pyomo expression is always the same depending on what started as a pyomo expression. If the user, chose something other than alpha/driver to make expression variables, we would effectively receive a different expression type.

PaulTalbot-INL replied:
I'm interested to know what those 12 tests are passing in as a string; is this the case where a variable is used as an `alpha` or `driver`? I agree that passing is consistent with the old behavior; I would like to clarify these options in the code so it's easier to know what's happening here. Did the error message tell you what the string was, or just tell you the string didn't work?

yoshiurr-INL replied:
The error shown was:

TypeError: ufunc 'true_divide' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

After googling this, it is an error that sent from pandas when you're attempting to use a string for a int/float operation.

PaulTalbot-INL replied:
okay; lots of questions remain for this, but I'm fine with this part working as it did before.

yoshiurr-INL committed:
Updated pyomo flag name and condition format

yoshiurr-INL committed:
Forgot to change test name for pyomo

yoshiurr-INL committed:
Since the added test is failing for pyomo, added the python3 only option to test if this would work

yoshiurr-INL committed:
Noticed the current library does not include pyomo. Removed test until further notice.

PaulTalbot-INL commented:
@yoshiurr-INL for the test, you can use `required_libraries = 'pyomo'` to cause the system to skip the test if pyomo isn't detected in the python env. For example, for a test in RAVEN that uses the optional `imageio` library:
```
  [./votingRegressor_plotting]
    type = 'RavenFramework'
    input = 'votingRegressor_plotting.xml'
    csv = 'data/outVotingRegressor_plotting.csv'
    image = 'data/votingRegressor_scatter-scatter-scatter-scatter-scatter.png'
    required_libraries = 'imageio'
    rel_err = 0.1
  [../]
```

PaulTalbot-INL commented:
Any progress on the test for this?

yoshiurr-INL committed:
Added new test to solve pyomo expression.

PaulTalbot-INL commented on tests/PyomoTest.py:
```diff
@@ -0,0 +1,283 @@
+def build_econ_settings(cfs, life=5, dr=0.1, tax=0.21, infl=0.02184):
+  """
+    Constructs global settings for econ run
+    @ In, FIXME
```
These FIXME in the docstrings need to be filled

yoshiurr-INL replied:
Done

PaulTalbot-INL commented on tests/PyomoTest.py:
```diff
@@ -0,0 +1,283 @@
+
+
+def build_generator(size, lifetime, dispatch):
+  # make the TEAL component
```
docstrings should be added where missing

yoshiurr-INL replied:
Done

PaulTalbot-INL commented on tests/PyomoTest.py:
```diff
@@ -0,0 +1,283 @@
+
+  market = build_market(market_size, market_life, prices, market_dispatch)
+
+  # DEBUGG
```
this commented line can be removed, it was a placeholder

yoshiurr-INL replied:
Done

PaulTalbot-INL commented on tests/PyomoTest.py:
```diff
@@ -0,0 +1,283 @@
+  # Objective: NPV
+  m.npv = pyo.Objective(expr=metrics['NPV'], sense=pyo.maximize)
+
+  print('DEBUGG PRE:')
```
how much of this do we want printing?

yoshiurr-INL replied:
We don't really need the pre-printing but post-printing I would like to keep

PaulTalbot-INL commented on tests/PyomoTest.py:
```diff
@@ -0,0 +1,283 @@
+  m.display()
+  print(m.gen_size.value)
+  calculated = pyo.value(m.npv) # Changed from print(m.npv.value)
+  correct = 10164.740285148146
```
I think we want to check several token dispatch optimized results, as well as the NPV. This will assure the price is correctly influencing the dispatch, and observe the "flipping" behavior

yoshiurr-INL replied:
Done

PaulTalbot-INL commented on tests/PyomoTest.py:
```diff
@@ -0,0 +1,283 @@
+    print('Success!')
+    sys.exit(0)
+  else:
+    print('ERROR: correct: {:1.3e}, calculated: {:1.3e}, diff {:1.3e}'.format(correct, calculated, correct-calculated))
```
printing the error actually won't cause the test to error out; you need a `sys.exit(1)` to register as a failed test

yoshiurr-INL replied:
Done

PaulTalbot-INL commented on tests/tests:
```diff
@@ -89,6 +89,20 @@
+  input = 'HourlyObjectOrientedTestWithPyomoExpression.py'
+  python3_only = True
+  required_libraries = 'pyomo'
+ [../]
```
I think this PyomoObjectOriented is superseded by the PyomoTest; also, I think it can be named just PyomoTest instead of CashFlow_PyomoTest

PaulTalbot-INL reviewed (changes requested):
a few cleanup changes needed still, plus test some dispatch points

yoshiurr-INL committed:
Updated PyomoTest as specified by comments and changed tests list to add minimum_library_versions for pyomo.
